tiledb/filter.py

```diff
@@ -477,6 +527,117 @@ def _attrs_(self):
         return {}
 
 
+class FloatScalingFilter(Filter):
+    """
+    Filter that stores floats as integers in a reduced representation via scaling.
+    The reduced storage space is in lieu of some precision loss. The float scaling
+    filter takes three parameters: the scale, the offset, and the byte width.
+
+    On write, the float scaling filter applies the scale factor and offset,
+    and stores the value of round((raw_float - offset) / scale) as an
+    integer with the specified NumPy dtype.
+
+    On read, the float scaling filter will reverse the scale factor and offset,
+    and returns the floating point data, with a potential loss of precision.
+
+    :param factor: the scaling factor used to translate the data
+    :type factor: float
+    :param offset: the offset value used to translate the data
+    :type offset: float
+    :param bytewidth: values may be stored as integers of bytewidth 1, 2, 4, or 8
+    :type np.integer:
+    :param ctx: A TileDB Context
+    :type ctx: tiledb.Ctx
+
+    **Example:**
+
+    >>> import tiledb, numpy as np, tempfile
+    >>> with tempfile.TemporaryDirectory() as tmp:
+    ...     dom = tiledb.Domain(tiledb.Dim(domain=(0, 9), tile=2, dtype=np.uint64))
+    ...     a1 = tiledb.Attr(name="a1", dtype=np.int64,
+    ...                      filters=tiledb.FilterList([tiledb.FloatScalingFilter(1, 0)]))
+    ...     schema = tiledb.ArraySchema(domain=dom, attrs=(a1,))
+    ...     tiledb.DenseArray.create(tmp + "/array", schema)
+
+    """
+
+    def __init__(
+        self,
+        factor: float = None,
+        offset: float = None,
+        bytewidth: int = None,
+        ctx: "Ctx" = None,
+    ):
+        # try:
+        #     self._factor = float(factor)
+        # except ValueError:
+        #     raise ValueError("`factor` argument must be a float")
+
+        # try:
+        #     self._offset = float(offset)
+        # except ValueError:
+        #     raise ValueError("`offset` argument must be a float")
+
+        # if bytewidth not in (1, 2, 4, 8):
+        #     raise ValueError("`bytewidth` argument must be 1, 2, 4, or 8")
+
+        self._factor = factor
+        self._offset = offset
+        self._bytewidth = bytewidth
+        self._ctx = ctx or default_ctx()
+
+        super().__init__(lt.FilterType.SCALE_FLOAT, self._ctx)
+
+        if factor:
+            self._set_option(
+                lt.Context(self._ctx.__capsule__(), False),
+                lt.FilterOption.SCALE_FLOAT_FACTOR,
+                float(factor),
```

Comment:
(BTW I will be cleaning up the code and adding better type checking for all this eventually.)